File: health_advanced_swift/HealthDashboard/Controllers/HDExerciseTimerViewController.m
//
//  HDExerciseTimerViewController.m
//  HealthDashboard
//
//  Created by zhang, haizi on 2026/3/21.
//

#import "HDExerciseTimerViewController.h"
#import "HDExerciseSummaryViewController.h"
#import "../Models/HDHealthDataModel.h"

@interface HDExerciseTimerViewController ()
@property (nonatomic, strong) UILabel *timeLabel;
@property (nonatomic, strong) UILabel *distanceLabel;
@property (nonatomic, strong) UILabel *paceLabel;
@property (nonatomic, strong) UILabel *calorieLabel;
@property (nonatomic, strong) UIProgressView *progressView;
@property (nonatomic, strong) UIButton *pauseButton;
@property (nonatomic, strong) UIButton *endButton;

@property (nonatomic, strong) NSTimer *timer;
@property (nonatomic, assign) NSInteger elapsedSeconds;
@property (nonatomic, assign) CGFloat currentDistance;
@property (nonatomic, assign) BOOL isRunning;
@end

@implementation HDExerciseTimerViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self setupUI];
    [self applyTheme];
    [self startTimer];
}

- (void)setupUI {
    self.view.backgroundColor = UIColor.systemBackgroundColor;
    self.title = @"运动中";
    
    // 时间显示
    _timeLabel = [UILabel new];
    _timeLabel.text = @"00:00:00";
    _timeLabel.font = [UIFont monospacedDigitSystemFontOfSize:48 weight:UIFontWeightBold];
    _timeLabel.textColor = UIColor.labelColor;
    _timeLabel.textAlignment = NSTextAlignmentCenter;
    [self.view addSubview:_timeLabel];
    
    // 距离显示
    _distanceLabel = [UILabel new];
    _distanceLabel.text = @"0.0 km";
    _distanceLabel.font = [UIFont monospacedDigitSystemFontOfSize:32 weight:UIFontWeightBold];
    _distanceLabel.textColor = UIColor.systemBlueColor;
    _distanceLabel.textAlignment = NSTextAlignmentCenter;
    [self.view addSubview:_distanceLabel];
    
    // 配速显示
    _paceLabel = [UILabel new];
    _paceLabel.text = @"配速: 0:00 /km";
    _paceLabel.font = [UIFont systemFontOfSize:16 weight:UIFontWeightRegular];
    _paceLabel.textColor = UIColor.secondaryLabelColor;
    _paceLabel.textAlignment = NSTextAlignmentCenter;
    [self.view addSubview:_paceLabel];
    
    // 卡路里显示
    _calorieLabel = [UILabel new];
    _calorieLabel.text = @"0 kcal";
    _calorieLabel.font = [UIFont systemFontOfSize:16 weight:UIFontWeightRegular];
    _calorieLabel.textColor = UIColor.secondaryLabelColor;
    _calorieLabel.textAlignment = NSTextAlignmentCenter;
    [self.view addSubview:_calorieLabel];
    
    // 进度条
    _progressView = [UIProgressView new];
    _progressView.progressTintColor = UIColor.systemBlueColor;
    _progressView.trackTintColor = UIColor.tertiaryLabelColor;
    [self.view addSubview:_progressView];
    
    // 暂停按钮
    _pauseButton = [UIButton buttonWithType:UIButtonTypeSystem];
    _pauseButton.titleLabel.font = [UIFont systemFontOfSize:16 weight:UIFontWeightSemibold];
    [_pauseButton setTitle:@"暂停" forState:UIControlStateNormal];
    _pauseButton.layer.cornerRadius = 8;
    _pauseButton.backgroundColor = UIColor.systemOrangeColor;
    [_pauseButton setTitleColor:UIColor.whiteColor forState:UIControlStateNormal];
    [_pauseButton addTarget:self action:@selector(togglePause) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:_pauseButton];
    
    // 结束按钮
    _endButton = [UIButton buttonWithType:UIButtonTypeSystem];
    _endButton.titleLabel.font = [UIFont systemFontOfSize:16 weight:UIFontWeightSemibold];
    [_endButton setTitle:@"结束" forState:UIControlStateNormal];
    _endButton.layer.cornerRadius = 8;
    _endButton.backgroundColor = UIColor.systemRedColor;
    [_endButton setTitleColor:UIColor.whiteColor forState:UIControlStateNormal];
    [_endButton addTarget:self action:@selector(endExercise) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:_endButton];
    
    [self layoutUI];
}

- (void)layoutUI {
    CGFloat padding = 16;
    CGFloat width = self.view.bounds.size.width - padding * 2;
    
    _timeLabel.frame = CGRectMake(padding, 100, width, 60);
    _distanceLabel.frame = CGRectMake(padding, 170, width, 50);
    _paceLabel.frame = CGRectMake(padding, 230, width, 24);
    _calorieLabel.frame = CGRectMake(padding, 260, width, 24);
    
    _progressView.frame = CGRectMake(padding, 300, width, 4);
    
    _pauseButton.frame = CGRectMake(padding, self.view.bounds.size.height - padding - 120, width / 2 - 8, 50);
    _endButton.frame = CGRectMake(padding + width / 2 + 8, self.view.bounds.size.height - padding - 120, width / 2 - 8, 50);
}

- (void)startTimer {
    _isRunning = YES;
    _elapsedSeconds = 0;
    _currentDistance = 0;
    
    _timer = [NSTimer scheduledTimerWithTimeInterval:1.0 target:self selector:@selector(updateTimer) userInfo:nil repeats:YES];
}

- (void)updateTimer {
    _elapsedSeconds++;
    _currentDistance += 0.01; // 模拟距离增加
    
    // 更新时间显示
    NSInteger hours = _elapsedSeconds / 3600;
    NSInteger minutes = (_elapsedSeconds % 3600) / 60;
    NSInteger seconds = _elapsedSeconds % 60;
    _timeLabel.text = [NSString stringWithFormat:@"%02ld:%02ld:%02ld", hours, minutes, seconds];
    
    // 更新距离显示
    _distanceLabel.text = [NSString stringWithFormat:@"%.1f km", _currentDistance];
    
    // 更新配速
    if (_currentDistance > 0) {
        CGFloat paceMinutes = _elapsedSeconds / 60.0 / _currentDistance;
        NSInteger paceMin = (NSInteger)paceMinutes;
        NSInteger paceSec = (NSInteger)((paceMinutes - paceMin) * 60);
        _paceLabel.text = [NSString stringWithFormat:@"配速: %ld:%02ld /km", paceMin, paceSec];
    }
    
    // 更新卡路里（约0.1 kcal/秒）
    NSInteger calories = (NSInteger)(_elapsedSeconds * 0.1);
    _calorieLabel.text = [NSString stringWithFormat:@"%ld kcal", calories];
    
    // 更新进度条（目标跑）
    if (_exerciseType == 0) {
        HDHealthDataModel *model = [HDHealthDataModel shared];
        CGFloat targetDistance = model.targetRunDistanceKM;
        _progressView.progress = MIN(1.0, _currentDistance / targetDistance);
    }
}

- (void)togglePause {
    if (_isRunning) {
        [_timer invalidate];
        _timer = nil;
        _isRunning = NO;
        [_pauseButton setTitle:@"继续" forState:UIControlStateNormal];
    } else {
        _isRunning = YES;
        _timer = [NSTimer scheduledTimerWithTimeInterval:1.0 target:self selector:@selector(updateTimer) userInfo:nil repeats:YES];
        [_pauseButton setTitle:@"暂停" forState:UIControlStateNormal];
    }
}

- (void)endExercise {
    [_timer invalidate];
    _timer = nil;
    
    // 创建运动记录
    HDExerciseRecord *record = [HDExerciseRecord new];
    record.type = _exerciseType;
    record.durationSeconds = _elapsedSeconds;
    record.distanceKM = _currentDistance;
    record.caloriesBurned = (NSInteger)(_elapsedSeconds * 0.1);
    record.timestamp = [NSDate date];
    
    // 保存到 Model
    [[HDHealthDataModel shared] saveExerciseRecord:record];
    
    // 跳转到总结页面
    HDExerciseSummaryViewController *vc = [HDExerciseSummaryViewController new];
    vc.exerciseRecord = record;
    [self.navigationController pushViewController:vc animated:YES];
}

- (void)applyTheme {
    self.view.backgroundColor = UIColor.systemBackgroundColor;
    _timeLabel.textColor = UIColor.labelColor;
    _distanceLabel.textColor = UIColor.systemBlueColor;
    _paceLabel.textColor = UIColor.secondaryLabelColor;
    _calorieLabel.textColor = UIColor.secondaryLabelColor;
}

- (void)dealloc {
    [_timer invalidate];
}

@end

File: health_advanced_swift/HealthDashboard/Controllers/HDExerciseTimerViewController.swift
//
//  HDExerciseTimerViewController.swift
//  HealthDashboard - 运动计时
//
//  Created by AI Assistant on 2026/3/22.
//

import UIKit

// MARK: - ViewController

@MainActor
class HDExerciseTimerViewController: UIViewController {

    // MARK: - Properties

    /// 运动类型（0=目标跑, 1=自由跑）
    var exerciseType: Int = 0

    private let timeLabel = UILabel()
    private let distanceLabel = UILabel()
    private let paceLabel = UILabel()
    private let calorieLabel = UILabel()
    private let progressView = UIProgressView()
    private let pauseButton = UIButton(type: .custom)
    private let endButton = UIButton(type: .custom)

    private var timer: Timer?
    private var elapsedSeconds: Int = 0
    private var currentDistance: CGFloat = 0
    private var isRunning: Bool = false

    // MARK: - Lifecycle

    override func viewDidLoad() {
        super.viewDidLoad()
        setupUI()
        applyTheme()
        startTimer()
        NotificationCenter.default.addObserver(self, selector: #selector(applyTheme), name: NSNotification.Name("HDThemeDidChange"), object: nil)
    }

    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        layoutUI()
    }

    // timer 在 endExercise/togglePause 中已 invalidate，VC 销毁时 Timer 自动释放

    // MARK: - UI Setup

    private func setupUI() {
        title = "运动中"

        timeLabel.text = "00:00:00"
        timeLabel.font = .monospacedDigitSystemFont(ofSize: 48, weight: .bold)
        timeLabel.textAlignment = .center
        view.addSubview(timeLabel)

        distanceLabel.text = "0.0 km"
        distanceLabel.font = .monospacedDigitSystemFont(ofSize: 32, weight: .bold)
        distanceLabel.textColor = .systemBlue
        distanceLabel.textAlignment = .center
        view.addSubview(distanceLabel)

        paceLabel.text = "配速: 0:00 /km"
        paceLabel.font = .systemFont(ofSize: 16)
        paceLabel.textColor = .secondaryLabel
        paceLabel.textAlignment = .center
        view.addSubview(paceLabel)

        calorieLabel.text = "0 kcal"
        calorieLabel.font = .systemFont(ofSize: 16)
        calorieLabel.textColor = .secondaryLabel
        calorieLabel.textAlignment = .center
        view.addSubview(calorieLabel)

        progressView.progressTintColor = .systemBlue
        progressView.trackTintColor = .tertiaryLabel
        view.addSubview(progressView)

        pauseButton.setTitle("暂停", for: .normal)
        pauseButton.titleLabel?.font = .systemFont(ofSize: 16, weight: .semibold)
        pauseButton.layer.cornerRadius = 8
        pauseButton.backgroundColor = .systemOrange
        pauseButton.setTitleColor(.white, for: .normal)
        pauseButton.addTarget(self, action: #selector(togglePause), for: .touchUpInside)
        view.addSubview(pauseButton)

        endButton.setTitle("结束", for: .normal)
        endButton.titleLabel?.font = .systemFont(ofSize: 16, weight: .semibold)
        endButton.layer.cornerRadius = 8
        endButton.backgroundColor = .systemRed
        endButton.setTitleColor(.white, for: .normal)
        endButton.addTarget(self, action: #selector(endExercise), for: .touchUpInside)
        view.addSubview(endButton)
    }

    private func layoutUI() {
        let padding: CGFloat = 16
        let width = view.bounds.width - padding * 2

        timeLabel.frame = CGRect(x: padding, y: 100, width: width, height: 60)
        distanceLabel.frame = CGRect(x: padding, y: 170, width: width, height: 50)
        paceLabel.frame = CGRect(x: padding, y: 230, width: width, height: 24)
        calorieLabel.frame = CGRect(x: padding, y: 260, width: width, height: 24)
        progressView.frame = CGRect(x: padding, y: 300, width: width, height: 4)

        let btnY = view.bounds.height - padding - 120
        let btnW = (width - 16) / 2
        pauseButton.frame = CGRect(x: padding, y: btnY, width: btnW, height: 50)
        endButton.frame = CGRect(x: padding + btnW + 16, y: btnY, width: btnW, height: 50)
    }

    // MARK: - Timer

    private func startTimer() {
        isRunning = true
        elapsedSeconds = 0
        currentDistance = 0
        /// NSTimer 在主线程 RunLoop 上触发，不阻塞主线程
        timer = Timer.scheduledTimer(timeInterval: 1.0, target: self, selector: #selector(updateTimer), userInfo: nil, repeats: true)
    }

    @objc private func updateTimer() {
        elapsedSeconds += 1
        currentDistance += 0.01

        // 更新时间
        let h = elapsedSeconds / 3600
        let m = (elapsedSeconds % 3600) / 60
        let s = elapsedSeconds % 60
        timeLabel.text = String(format: "%02d:%02d:%02d", h, m, s)

        // 更新距离
        distanceLabel.text = String(format: "%.1f km", currentDistance)

        // 更新配速
        if currentDistance > 0 {
            let paceMinutes = Double(elapsedSeconds) / 60.0 / Double(currentDistance)
            let paceMin = Int(paceMinutes)
            let paceSec = Int((paceMinutes - Double(paceMin)) * 60)
            paceLabel.text = String(format: "配速: %d:%02d /km", paceMin, paceSec)
        }

        // 更新卡路里（约0.1 kcal/秒）
        calorieLabel.text = "\(Int(Double(elapsedSeconds) * 0.1)) kcal"

        // 更新进度条（仅目标跑）
        if exerciseType == 0 {
            let model = HDHealthDataModel.shared()
            let targetDist = CGFloat(model.targetRunDistanceKM)
            if targetDist > 0 {
                progressView.progress = Float(min(1.0, currentDistance / targetDist))
            }
        }
    }

    // MARK: - Actions

    @objc private func togglePause() {
        if isRunning {
            timer?.invalidate()
            timer = nil
            isRunning = false
            pauseButton.setTitle("继续", for: .normal)
        } else {
            isRunning = true
            timer = Timer.scheduledTimer(timeInterval: 1.0, target: self, selector: #selector(updateTimer), userInfo: nil, repeats: true)
            pauseButton.setTitle("暂停", for: .normal)
        }
    }

    @objc private func endExercise() {
        timer?.invalidate()
        timer = nil

        // 构建运动记录
        let record = HDExerciseRecord()
        record.type = exerciseType
        record.durationSeconds = elapsedSeconds
        record.distanceKM = currentDistance
        record.caloriesBurned = Int(Double(elapsedSeconds) * 0.1)
        record.timestamp = Date()

        // 通过 Model 方法保存数据
        HDHealthDataModel.shared().saveExerciseRecord(record)

        // 跳转到总结页
        let vc = HDExerciseSummaryViewController()
        vc.exerciseRecord = record
        navigationController?.pushViewController(vc, animated: true)
    }

    // MARK: - Theme

    @objc func applyTheme() {
        view.backgroundColor = .systemBackground
        timeLabel.textColor = .label
        distanceLabel.textColor = .systemBlue
        paceLabel.textColor = .secondaryLabel
        calorieLabel.textColor = .secondaryLabel
    }
}
